src/server/chromium/crDragDrop.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+import { CRPage } from './crPage';
+import * as types from '../types';
+import { assert } from '../../utils/utils';
+import { Protocol } from './protocol';
+import { toModifiersMask } from './crProtocolHelper';
+
+export class DragManager {
+  private _page: CRPage;
+  private _dragState: Protocol.Input.DragData | null = null;
+  constructor(page: CRPage) {
+    this._page = page;
+  }
+
+  async cancelDrag() {
+    if (!this._dragState)
+      return false;
+    await this._page._mainFrameSession._client.send('Input.dispatchDragEvent', {
+      type: 'dragCancel',
+      x: 0,
+      y: 0,
+      data: {
+        items: [],
+        dragOperationsMask: 0xFFFF,
+      }
+    });
+    this._dragState = null;
+    return true;
+  }
+
+  async interceptDragCausedByMove(x: number, y: number, button: types.MouseButton | 'none', buttons: Set<types.MouseButton>, modifiers: Set<types.KeyboardModifier>, moveCallback: () => Promise<void>): Promise<void> {
+    if (this._dragState) {
+      await this._page._mainFrameSession._client.send('Input.dispatchDragEvent', {
+        type: 'dragOver',
+        x,
+        y,
+        data: this._dragState,
+        modifiers: toModifiersMask(modifiers),
+      });
+      return;
+    }
+    if (button !== 'left')
+      return moveCallback();
+
+    const client = this._page._mainFrameSession._client;
+    let onDragIntercepted: (payload: Protocol.Input.dragInterceptedPayload) => void;
+    const dragInterceptedPromise = new Promise<Protocol.Input.dragInterceptedPayload>(x => onDragIntercepted = x);
+
+    const dragSignalPromises: Promise<boolean>[] = [];
+    for (const frame of this._page._page.frames()) {
+      const promise = frame.evaluateExpression((function(){
+        return new Promise<boolean>(callback => {
+          document.addEventListener('mousemove', () => setTimeout(() => callback(false), 0), {once: true});
+          document.addEventListener('dragstart', () => callback(true), {once: true});
+        });
+      }).toString(), true, undefined, 'utility');
+      // don't resolve on errors. This could be replaced by Promise.any after Node v15
+      dragSignalPromises.push(promise.catch(e => new Promise(() => {})));
```

Comment:
Why not resolve on errors? What if we have just a single frame and is has just navigated under our feet? We'll stall forever.

Reply:
It doesn't stall because these are used with `Promise.race`